# DS 3000 - Assignment 2

**Student Name**: SHREYAS SHUKLA

**Date**: 14 SEP 2024


### Submission Instructions
Submit this `ipynb` file to canvas.

The `ipynb` format stores outputs from the last time you ran the notebook.  (When you open a notebook it has the figures and outputs of the last time you ran it too).  To ensure that your submitted `ipynb` file represents your latest code, make sure to give a fresh run `Kernel > Restart & Run All` just before uploading the `ipynb` file to Canvas.

### Academic Integrity

**Writing your homework is an individual effort.**  You may discuss general python problems with other students but under no circumstances should you observe another student's code which was written for this assignment, from this year or past years.  Pop into office hours or DM us in MS Teams if you have a specific question about your work or if you would like another pair of eyes or talk through your code.

Don't forget to cite websites which helped you solve a problem in a unique way.  You can do this in markdown near the code or with a simple one-line comment. You do not need to cite the official python documentation.

**Documentation / style counts for credit**  Please refer to the Pep-8 style, to improve the readability and consistency of your Python code. For more information, read the following article [How to Write Beautiful Python Code With PEP 8](https://realpython.com/python-pep8/) or ask your TA's for tips.

**NOTE:<span style='color:red'> Write python expressions to answer ALL questions below and ensure that you use the `print()` function to display the output.</span>** Each question should be answered in a new code cell. For example, your solution for question 1.1 should be in a different code cell from your solution for question 1.2.

## Question 1: 

Question 1: Analyzing numeric Data (50 pts)

The data that you are working with pertains to the risk factors that are associated with low infant birth weight**. The data contains the following fields:

low
indicator of birth weight less than 2.5 kg.

age
mother's age in years.

lwt
mother's weight in pounds at last menstrual period.

race
mother's race (1 = white, 2 = black, 3 = other).

smoke
smoking status during pregnancy  (0 = false, 1 = true).

ptl
number of previous premature labors.

ht
history of hypertension (0 = false, 1 = true).

ui
presence of uterine irritability.

ftv
number of physician visits during the first trimester.

bwt
birth weight in grams.

Write python code to answer the questions below and ensure that you round all numeric calculations (i.e. floats) to 2 decimal places (unless indicated otherwise).
Tip: you will need to filter for conditions that are true before performing certain calculations.

(5 pts) Load the attached data into a numpy array and display the dimensions of the data: birthwt.csvDownload birthwt.csv
(3 pts) What is the age range of mothers in this dataset.
Note: the range should include: the minimum, average and maximum age. Ensure that the ages are represented as whole numbers.
(2 pts) How many mothers smoked during pregnancy?
(2 pts) How many mothers had a history of hypertension?
(3 pts) How many mothers have 1 or more premature births?
(5 pts) Create a function called convert_birthweight() that: 1) takes grams as input, 2) converts grams to pounds and ounces, and 3) return the converted value. Use the convert_birthweight() function to convert the birth weight to pounds and ounces in the questions 7-11 below.
(5 pts) What is the average birth weight of babies whose mothers have a history of premature birth? 
(5 pts) What is the minimum, average and maximum birth weight of babies whose mother did not smoke during pregnancy and do not have a history of hypertension?
(5 pts) What is the minimum, average and maximum birth weight of babies whose mothers smoked during pregnancy?
(5 pts) What is the minimum, average and maximum birth weight of babies whose mothers had a history of hypertension? 
(5 pts) What is the minimum, average and maximum birth weight of babies whose mothers smoked during pregnancy and had a history of hypertension? 
(5 pts) Compare and contrast the results from 7-11 above and summarize your findings.
Do you see any patterns in the birth weight of babies whose mother did not have any history of hypertension and/or did not smoke during pregnancy versus those who did? Ensure that you reference specific birth weights to support your explanation.


In [7]:
import numpy as np

# Load data from CSV file
data = np.genfromtxt('birthwt.csv', delimiter=',', skip_header=1)
print(f"Data dimensions: {data.shape}")

# Calculate the age range of mothers
ages = data[:, 1]
min_age = int(np.min(ages))
avg_age = int(np.mean(ages))
max_age = int(np.max(ages))
print(f"Mother's age range: Min = {min_age}, Avg = {avg_age}, Max = {max_age}")

# Count the number of mothers who smoked during pregnancy
smokers_count = np.sum(data[:, 4] == 1)
print(f"Mothers who smoked during pregnancy: {smokers_count}")

# Count the number of mothers with a history of hypertension
hypertension_count = np.sum(data[:, 6] == 1)
print(f"Mothers with hypertension: {hypertension_count}")

# Count the number of mothers who had one or more premature births
premature_births_count = np.sum(data[:, 5] >= 1)
print(f"Mothers with 1 or more premature births: {premature_births_count}")

# Convert birth weight from grams to pounds and ounces
def convert_birthweight(grams):
    pounds = grams / 453.59237
    ounces = (pounds - int(pounds)) * 16
    return int(pounds), round(ounces, 2)

# Calculate the average birth weight for babies of mothers with premature births
premature_birth_weights = data[data[:, 5] >= 1][:, 9]
avg_premature_weight = np.mean(premature_birth_weights)
print(f"Avg birth weight for premature birth cases: {avg_premature_weight:.2f} grams")

# Calculate birth weight statistics for babies of non-smokers without hypertension
non_smokers_no_htn = data[(data[:, 4] == 0) & (data[:, 6] == 0)][:, 9]
min_bwt_ns_nh = np.min(non_smokers_no_htn)
avg_bwt_ns_nh = np.mean(non_smokers_no_htn)
max_bwt_ns_nh = np.max(non_smokers_no_htn)
print(f"Birth weight (non-smokers, no hypertension): Min = {min_bwt_ns_nh:.2f}, "
      f"Avg = {avg_bwt_ns_nh:.2f}, Max = {max_bwt_ns_nh:.2f} grams")

# Birth weight stats for babies of mothers who smoked during pregnancy
smokers_bwt = data[data[:, 4] == 1][:, 9]
min_bwt_smokers = np.min(smokers_bwt)
avg_bwt_smokers = np.mean(smokers_bwt)
max_bwt_smokers = np.max(smokers_bwt)
print(f"Birth weight (smokers): Min = {min_bwt_smokers:.2f}, "
      f"Avg = {avg_bwt_smokers:.2f}, Max = {max_bwt_smokers:.2f} grams")

# Birth weight stats for babies of mothers with hypertension
hypertension_bwt = data[data[:, 6] == 1][:, 9]
min_bwt_htn = np.min(hypertension_bwt)
avg_bwt_htn = np.mean(hypertension_bwt)
max_bwt_htn = np.max(hypertension_bwt)
print(f"Birth weight (hypertension): Min = {min_bwt_htn:.2f}, "
      f"Avg = {avg_bwt_htn:.2f}, Max = {max_bwt_htn:.2f} grams")

# Birth weight stats for babies of mothers who smoked and had hypertension
smokers_htn_bwt = data[(data[:, 4] == 1) & (data[:, 6] == 1)][:, 9]
min_bwt_smokers_htn = np.min(smokers_htn_bwt)
avg_bwt_smokers_htn = np.mean(smokers_htn_bwt)
max_bwt_smokers_htn = np.max(smokers_htn_bwt)
print(f"Birth weight (smokers & hypertension): Min = {min_bwt_smokers_htn:.2f}, "
      f"Avg = {avg_bwt_smokers_htn:.2f}, Max = {max_bwt_smokers_htn:.2f} grams")


Data dimensions: (189, 10)
Mother's age range: Min = 14, Avg = 23, Max = 45
Mothers who smoked during pregnancy: 74
Mothers with hypertension: 12
Mothers with 1 or more premature births: 30
Avg birth weight for premature birth cases: 2579.40 grams
Birth weight (non-smokers, no hypertension): Min = 1021.00, Avg = 3090.44, Max = 4990.00 grams
Birth weight (smokers): Min = 709.00, Avg = 2771.92, Max = 4238.00 grams
Birth weight (hypertension): Min = 1135.00, Avg = 2536.83, Max = 3790.00 grams
Birth weight (smokers & hypertension): Min = 1135.00, Avg = 2561.00, Max = 3756.00 grams


## Part 12) 

The data clearly demonstrates a pattern linking lower birth weights to both hypertension and smoking. The moms who did not smoke or have a history of hypertension gave birth to babies with the highest average birth weight—3090.44 grams—with the lowest weight at 1021.00 grams and the highest weight at 4990.00 grams. On the other hand, kids born to smokers had an average birth weight of 2771.92 grams, with a range of 709.00 to 4238.00 grams. The average birth weight of babies born to moms with hypertension was 2536.83 grams, with a range of 1135.00 to 3790.00 grams. With an average birth weight of 2561.00 grams, babies delivered to moms who smoked and had hypertension were the most affected group. According to these results, birth weight is impacted negatively by smoking and hypertension, and birth weights that are healthy are linked to the absence of these risk factors.